conmon-rs/server/src/fd_socket.rs

```diff
@@ -0,0 +1,268 @@
+//! ...
+//! u64: slot for file descriptor num_fds-1
+//! ```
+//!
+//! On error `num_fds` will be set to 0xffff_ffff and the rest of the message
+//! is an error message string.
+//! ```txt
+//! u64: (request_id << 32) | 0xffff_ffff
+//! ...: error message
+//! ```
+//!
+//! **Close file descriptors:**
+//! Received file descriptors will be closed on disconnect
+//! or with an empty request (`request_id` = 0 and `num_fds` = 0).
+//!
+//! An empty request does **not** receive a response.
+
+use crate::listener::{Listener, SeqpacketListener};
+use anyhow::format_err;
+use anyhow::Result;
+use std::{
+    collections::{hash_map, HashMap},
+    io::{IoSliceMut, Write},
+    mem,
+    num::Wrapping,
+    os::fd::OwnedFd,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+use tokio::{runtime::Handle, sync::Mutex as AsyncMutex, task};
+use tokio_seqpacket::{ancillary::OwnedAncillaryMessage, UnixSeqpacket};
+use tracing::{debug_span, Instrument};
+
+macro_rules! lock {
+    ($x:expr) => {
+        $x.lock().map_err(|e| format_err!("{:#}", e))?
+    };
+}
+
+#[derive(Debug, Default)]
+pub struct FdSocket {
+    server: AsyncMutex<Option<Server>>,
+    state: Mutex<State>,
+}
+
+impl FdSocket {
+    pub async fn start(self: Arc<Self>, path: PathBuf) -> Result<PathBuf> {
+        let mut server = self.server.lock().await;
+        if let Some(server) = server.as_ref() {
+            Ok(server.path.clone())
+        } else {
+            *server = Server::start(path.clone(), self.clone()).await?.into();
+            Ok(path)
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn take(&self, slot: u64) -> Result<OwnedFd> {
+        lock!(self.state).take(slot)
+    }
+
+    pub fn take_all<I>(&self, slots: I) -> Result<Vec<OwnedFd>>
+    where
+        I: IntoIterator<Item = u64>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        let slots = slots.into_iter();
+        if slots.len() == 0 {
+            Ok(Vec::new())
+        } else {
+            let mut state = lock!(self.state);
+            slots.into_iter().map(|slot| state.take(slot)).collect()
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct State {
+    last: Wrapping<u64>,
+    fds: HashMap<u64, OwnedFd>,
+}
+
+impl State {
+    fn add(&mut self, fd: OwnedFd) -> u64 {
+        let mut slot = self.last;
+        loop {
+            slot += 1;
+            match self.fds.entry(slot.0) {
+                hash_map::Entry::Occupied(_) => continue,
+                hash_map::Entry::Vacant(entry) => {
+                    entry.insert(fd);
+                    break;
+                }
+            }
+        }
+        println!("add {slot}: {:?}", self.fds);
+        self.last = slot;
+        slot.0
+    }
+
+    fn take(&mut self, slot: u64) -> Result<OwnedFd> {
+        println!("take {slot}: {:?}", self.fds);
+        self.fds
+            .remove(&slot)
+            .ok_or_else(|| anyhow::anyhow!("no file descriptor in slot {slot}"))
+    }
+}
+
+#[derive(Debug, Default)]
```

Comment:
I don't think we need a default implementation for `Server`:
```suggestion
#[derive(Debug)]
```